agent/src/softwarecontaineragent.cpp

```diff
@@ -132,26 +133,34 @@ ContainerID SoftwareContainerAgent::findSuitableId()
     return availableID;
 }
 
-ContainerID SoftwareContainerAgent::createContainer(const std::string &prefix, const std::string &config)
+SoftwareContainer* SoftwareContainerAgent::makeSoftwareContainer(const ContainerID &containerID)
+{
+    log_debug() << "Created container with ID :" << containerID;
+    auto container = new SoftwareContainer(m_softwarecontainerWorkspace, "SC-" + std::to_string(containerID));
+    return container;
+}
+
+ContainerID SoftwareContainerAgent::createContainer(const std::string &config)
 {
     profilepoint("createContainerStart");
     profilefunction("createContainerFunction");
 
     parseConfig(config);
 
     SoftwareContainer *container;
-    if (m_preloadedContainers.size() != 0) {
-        container = m_preloadedContainers[0].release();
-        m_preloadedContainers.erase(m_preloadedContainers.begin());
+    ContainerID availableID;
+    if (!m_preloadedContainers.empty()) {
+        auto pair = std::move(m_preloadedContainers.front());
+        m_preloadedContainers.pop();
+
+        availableID = pair.first;
+        container   = pair.second.release();
     } else {
-        container = new SoftwareContainer(m_softwarecontainerWorkspace);
+        availableID = findSuitableId();
+        container = makeSoftwareContainer(availableID);
```

Comment:
Since we only want to modify the container below, and only if it's not already preloaded (then init() is superfluous), can't we use insert on m_containers with a pair in both clauses somehow? I think it would be worth a shot. If it can be easily simplified, I would like to see that. If not, then nevermind.